<div class="alert alert-success">
<h2> Комментарий ревьюера</h2>

Сергей, привет! Спасибо, что прислал работу во время :) 
    
Меня зовут Евгений, и я буду проверять твой проект) 
Ты проделал большую работу над проектом, однако в любом деле всегда можно можно что-то улучшить. Моя основная цель — не показать, что где—то совершены какие—либо ошибки, а поделиться своим опытом, который поможет тебе в дальнейших работах. Далее в файле ты сможешь увидить мои комментарии, постарайся учесть их при выполнении следующих заданий.    

Мои комментарии обозначены пометкой **Комментарий ревьюера**. Далее в файле ты сможешь найти их в похожих ячейках (если фон комментария <font color=green>зелёный</font> - всё сделано правильно; <font color=orange>оранжевый</font> - есть замечания, но не критично; <font color=red>красный</font> - нужно переделать). Не удаляй эти комментарии и постарайся учесть их в ходе выполнения данного проекта. 

</div>

## Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<div class="alert alert-success">
<h2> Комментарий ревьюера</h2>

Вступление в работу очень важно. Так ты поясняешь то, чему она посвящена. Можно также привести краткий план проведения самой работы и описание используемых данных. 

</div>

### Шаг 1. Откройте файл с данными и изучите общую информацию. 

In [1]:
import pandas as pd
df = pd.read_csv('/datasets/data.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


### Вывод

Мы видим, что перед нами датафрейм с 21525 строчками и 12 столбцами. Имена столбцов записаны корректно. Типа некоторых данных, возможно, придётся изменить.

### Шаг 2. Предобработка данных

### Обработка пропусков

In [2]:
#узнаем, сколько всего пропусков в данных
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [3]:
#визуально ознакомимся с таблицей
df.head(50)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   
20         1   -1311.604166         36              среднее             1   
21         1    -253.685166         33              среднее             1   
22         1   -1766.644138         24              среднее             1   
23         0    -272.981385         21               высшее             0   
24         1  338551.952911         57              среднее             1   
25         0  363548.489348         67              среднее             1   
26         0            NaN         41              среднее             1   
27         0    -529.191635         28               высшее             0   
28         1    -717.274324         26               высшее             0   
29         0            NaN         63              среднее             1   
30         1  335581.668515         62              среднее             1   
31         0   -1682.083438         47            начальное             3   
32         0   -4649.910832         34              СРЕДНЕЕ             1   
33         0   -1548.637544         48              среднее             1   
34         0   -4488.067031         35              среднее             1   
35         0  394021.072184         68              среднее             1   
36         0    -176.216688         33              среднее             1   
37         0   -6448.810860         43               ВЫСШЕЕ             0   
38         0    -597.881827         25               высшее             0   
39         0    -650.587796         31               высшее             0   
40         0   -1030.362606         30              среднее             1   
41         0            NaN         50              среднее             1   
42         0   -1257.496190         20  неоконченное высшее             2   
43         0   -4375.681384         43  неоконченное высшее             2   
44         1   -1362.041728         26              СРЕДНЕЕ             1   
45         1   -1039.451741         49              среднее             1   
46         2   -2262.712304         37              среднее             1   
47         1   -2689.137274         33               высшее             0   
48         0   -3341.067886         45              среднее             1   
49         0   -1181.443228         54              среднее             1   

            family_status  family_status_id gender  income_type  debt  \
0  

In [4]:
#посмотрим, какие типы дохода есть в данных
unique_income_type = df['income_type'].unique()
unique_income_type

array(['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете'],
      dtype=object)

In [5]:
#узнаем, сколько отсутствующих значений в каждом типе, чтобы потом прописать замену только для них
#df['nan_income'] = df['total_income'].isna()
#df.groupby('income_type').agg({'total_income': 'count', 'nan_income': 'sum'}).astype(int)


In [6]:
#заменим пустые значения в этих типах дохода медианными
gov_median = df.loc[df['income_type'] == 'госслужащий', 'total_income'].median()
df.loc[df['income_type'] == 'госслужащий', 'total_income'] = df.loc[df['income_type'] == 'госслужащий', 'total_income'].fillna(gov_median)

companion_median = df.loc[df['income_type'] == 'компаньон', 'total_income'].median()
df.loc[df['income_type'] == 'компаньон', 'total_income'] = df.loc[df['income_type'] == 'компаньон', 'total_income'].fillna(companion_median)

pension_median = df.loc[df['income_type'] == 'пенсионер', 'total_income'].median()
df.loc[df['income_type'] == 'пенсионер', 'total_income'] = df.loc[df['income_type'] == 'пенсионер', 'total_income'].fillna(pension_median)

business_median = df.loc[df['income_type'] == 'предприниматель', 'total_income'].median()
df.loc[df['income_type'] == 'предприниматель', 'total_income'] = df.loc[df['income_type'] == 'предприниматель', 'total_income'].fillna(business_median)

employee_median = df.loc[df['income_type'] == 'сотрудник', 'total_income'].median()
df.loc[df['income_type'] == 'сотрудник', 'total_income'] = df.loc[df['income_type'] == 'сотрудник', 'total_income'].fillna(employee_median)

# проверим, остались ли пропуски
df['nan_income'] = df['total_income'].isna()
df.groupby('income_type').agg({'total_income': 'count', 'nan_income': 'sum'}).astype(int)

total_income  nan_income
income_type                              
безработный                 2           0
в декрете                   1           0
госслужащий              1459           0
компаньон                5085           0
пенсионер                3856           0
предприниматель             2           0
сотрудник               11119           0
студент                     1           0


<div class="alert alert-success">
<h2> Комментарий ревьюера</h2>

Очень хорошо, что производим замену пропущенных значений с помощью выявленных групп.  
    
Давай попробуем ещё немного упростить с помощью методов `groupby()`, `trasform()` и `fillna()`.
    
Руководствоваться будем следующей логикой:
    
1. Создадим отдельный столбец, в который поместим медианное значение для каждого объекта исходя из группы профессий. В этом нам поможет метод `transform()`
2. Далее заменим пропуски используя полученный столбец
    
---
    
Хотел бы отметить, что это всего лишь один метод решения этой задачи.
    
</div>

In [7]:
# удалим не нужные для дальнейшего исследования столбцы days_employed и nan_income
del df['days_employed']
del df['nan_income']

### Вывод

Мы обнаружили по 2174 пропуска: в столбцах days_employed (стаж) и total_income (доход). Эти пропуски, вероятно, являются случайными. Рабочий стаж не имеет значения для нашего исследования, поэтому мы избавились от этого столбца. Пропуски в столбце с доходами мы заменили медианными значениями.

### Замена типа данных

In [8]:
#заменим значения размера дохода на целые
df['total_income'] = df['total_income'].astype(int)

### Вывод

Вещественные числа в столбце total_income мы заменили целыми. В округлённом виде зарплата читается лучше.

### Обработка дубликатов

In [9]:
# узнаем, сколько полных дубликатов в наших данных
df.duplicated().sum()

54

In [10]:
# приведём к единому регистру значения в столбцах education и family_status
df['education'] = df['education'].str.lower()
df['family_status'] = df['family_status'].str.lower()
# посмотрим, остались ли одинаковые значения в разных регистрах
print(df['education'].unique())
print(df['family_status'].unique())

['высшее' 'среднее' 'неоконченное высшее' 'начальное' 'ученая степень']
['женат / замужем' 'гражданский брак' 'вдовец / вдова' 'в разводе'
 'не женат / не замужем']


### Вывод

В столбце education значения были написаны в разных регистрах. Также в разных регистрах были написаны значения столбца family_status. Я привёл их все к нижнему регистру, чтобы позднее было удобнее их анализировать.

### Избавляемся от артефактов


Я обратил внимание, что в столбце children есть аномальные значения: -1 и 20. Можно предположить, что -1 это на самом деле 1 со случайно напечатанным минусом. Насчет цифры 20 может быть два варианта: 1) это 2 со случайно напечатанным нулём 2) у человека действительно 20 детей: например своих и приёмных. Поменяем -1 на 1 везде. Клиентов со значением 20 в строке children уберём из данных (они составляют всего 0,3% от всех клиентов).

In [11]:
# очистим данные от тех клиентов, которым приписывается наличие 20 детей
df = df[df['children'] != 20]
#Заменим значения -1 на 1
df.loc[df["children"] == -1,"children"] = 1
# проверим, получилось ли
df['children'].value_counts()

0    14149
1     4865
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

### Лемматизация

In [12]:
# найдем наиболее часто употребляемые в целях существительные - одновременно цели кредитования
from collections import Counter
from pymystem3 import Mystem
m = Mystem()
lemm_purposes = Counter(m.lemmatize(' '.join(df['purpose'])))
lemm_purposes

Counter({'покупка': 5893,
         ' ': 55001,
         'жилье': 4458,
         'приобретение': 462,
         'автомобиль': 4299,
         'дополнительный': 905,
         'образование': 4007,
         'сыграть': 770,
         'свадьба': 2339,
         'операция': 2600,
         'с': 2911,
         'на': 2222,
         'проведение': 773,
         'для': 1292,
         'семья': 640,
         'недвижимость': 6346,
         'коммерческий': 1308,
         'жилой': 1230,
         'строительство': 1874,
         'собственный': 629,
         'подержать': 849,
         'свой': 2231,
         'со': 629,
         'заниматься': 908,
         'сделка': 940,
         'получение': 1313,
         'высокий': 1369,
         'подержанный': 109,
         'профильный': 434,
         'сдача': 652,
         'ремонт': 610,
         '\n': 1})

In [13]:
# лемматизируем значения столбца purpose
df['purpose'] = df['purpose'].apply(m.lemmatize)
# напишем функцию, которая поможет сократить цель кредита до одного слова
def one_purpose(row):
    for purpose in row:
        if 'свадьба' in row:
            return 'свадьба'
        if 'образование' in row:
            return 'образование'
        if 'недвижимость' in row:
            return 'недвижимость'
        if 'жилье' in row:
            return 'недвижимость'
        if 'автомобиль' in row:
            return 'автомобиль'
        return 'прочее'
df['new_purpose'] = df['purpose'].apply(one_purpose)
df['new_purpose'].value_counts()

недвижимость    10804
автомобиль       4299
образование      4007
свадьба          2339
Name: new_purpose, dtype: int64


<div class="alert alert-success">
<h2> Комментарий ревьюера</h2>

Цели кредитов определены верно.     
</div>

### Вывод

Цели, на которые люди брали кредиты, в изначальных данных были написаны по-разному. Мы привели их к единому виду, так что теперь они указаны в одно слово. Всего я выделил четыре главных цели: недвижимость, образование, автомобиль и свадьба. Жилая и коммерческая недвижимость объединены. С целями, указанными в таком виде, нам будет удобнее отвечать на вопрос о том, зависит ли возврат кредита в срок от цели кредита.

### Категоризация данных

In [14]:
# создадим словарь, в который поместим столбцы с доходом и наличием задолженности
income_dict = df[['total_income','debt']]
# удалим из словаря дубликаты и сбросим индексы
income_dict = income_dict.drop_duplicates().reset_index(drop=True)
df_income_grouped = income_dict.groupby('total_income').count()
# Напишем функцию для разделения клиентов по размеру доходов
def income_group(total_income):
    if total_income <= 100000:
        return 'Доход до 100 тысяч рублей'
    if total_income <= 200000:
        return 'Доход 100-200 тысяч рублей'
    return 'Доход более 200 тысяч рублей'
# добавим в словарь столбец с названиями групп, на которые разбили клиентов
income_dict['income_groups'] = income_dict['total_income'].apply(income_group)
income_dict.head()

total_income  debt                 income_groups
0        253875     0  Доход более 200 тысяч рублей
1        112080     0    Доход 100-200 тысяч рублей
2        145885     0    Доход 100-200 тысяч рублей
3        267628     0  Доход более 200 тысяч рублей
4        158616     0    Доход 100-200 тысяч рублей


<div class="alert alert-warning">
<h2> Комментарий ревьюера</h2>

Лучше писать о том, почему выбрал данные категории. Так быстрее понять используемую логику в деталях проекта. 
    
</div>

In [15]:
# создадим словарь, в который поместим цель кредита, наличие долга и общий доход.
#Столбец с доходами пригодится, чтобы при создании сводной таблицы по этому столбцу считались строки
purpose_dict = df[['new_purpose','debt', 'total_income']]
# удалим из словаря дубликаты и сбросим индексы
purpose_dict = purpose_dict.drop_duplicates().reset_index(drop=True)

purpose_dict.head()

new_purpose  debt  total_income
0  недвижимость     0        253875
1    автомобиль     0        112080
2  недвижимость     0        145885
3   образование     0        267628
4       свадьба     0        158616

In [16]:
# создадим словарь, в который поместим столбцы с количеством детей, наличием долга и доходом 
children_dict = df[['children','debt', 'total_income']]
# удалим из словаря дубликаты и сбросим индексы
children_dict = children_dict.drop_duplicates().reset_index(drop=True)
df_children_grouped = children_dict.groupby('children').count()
# напишем функцию, которая поможет сгруппировать людей на тех, у кого дети есть и у кого их нет
def children_group(children):
    if children > 0:
        return 'Есть дети'
    return 'Детей нет'
children_dict['children_groups'] = children_dict['children'].apply(children_group)
children_dict.head()

children  debt  total_income children_groups
0         1     0        253875       Есть дети
1         1     0        112080       Есть дети
2         0     0        145885       Детей нет
3         3     0        267628       Есть дети
4         0     0        158616       Детей нет

In [17]:
# создадим словарь, в который поместим столбцы с семейным статусом, доходом и наличием долга
family_dict = df[['family_status','debt', 'total_income']]
# удалим из словаря дубликаты и сбросим индексы
family_dict = family_dict.drop_duplicates().reset_index(drop=True)

family_dict.head()

family_status  debt  total_income
0   женат / замужем     0        253875
1   женат / замужем     0        112080
2   женат / замужем     0        145885
3   женат / замужем     0        267628
4  гражданский брак     0        158616

### Вывод

Чтобы в следующем разделе ответить на четыре вопроса, мы создали четыре словаря. Теперь у нас есть отдельные датафреймы для тех, у кого есть дети и укого их нет, для людей с разным уровнем дохода, семейным статусом и целью кредита. В эти словари я также поместил данные о наличии задолженности и месячном доходе. Имея четыре таких словаря, мы сможем ответить на вопрос о зависимости просрочки от вышеперечисленных данных (наличие детей, семейный статус, цель кредита и уровень дохода).

Посколько нам нужно ответить на вопрос "Есть ли зависимость между наличием детей и возвратом кредита в срок?", то точное количество детей неважно. Поэтому я оставил только две группы: есть дети и детей нет. По размерам дохода я резделил клиентов на три более или менее равные группы.

### Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [18]:
# Сделаем сводную таблицу 
children_pivot = children_dict.pivot_table(index = ['children_groups'], 
                                columns = 'debt', 
                                values = 'total_income', 
                                aggfunc = 'count').fillna(0)
# у тех, кто имеет 5 детей, было NaN там, где должен быть 0,
#поэтому я заменил отсутствующие значения на 0.
new = ['Задолженность отсутствует', 'Была задолженность']
children_pivot.set_axis(new, axis = 'columns', inplace = True)
children_pivot['%'] = children_pivot['Была задолженность'] / children_pivot['Задолженность отсутствует'] * 100

children_pivot


Задолженность отсутствует  Была задолженность          %
children_groups                                                          
Детей нет                            11489                 955   8.312299
Есть дети                             5948                 620  10.423672

### Вывод

Зависимость есть. 8,3% из тех, у кого нет детей, имели просрочку по кредиту. Среди тех, у кого дети есть, просрочки встречаются чаще: таких здесь 10,4%.


<div class="alert alert-success">
<h2> Комментарий ревьюера</h2>

Верно.     
</div>

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [19]:
family_pivot = family_dict.pivot_table(index = ['family_status'], 
                                columns = 'debt', 
                                values = 'total_income', 
                                aggfunc = 'count')
new = ['Задолженность отсутствует', 'Была задолженность']
family_pivot.set_axis(new, axis = 'columns', inplace = True)
family_pivot['%'] = family_pivot['Была задолженность'] / family_pivot['Задолженность отсутствует'] * 100

family_pivot

Задолженность отсутствует  Была задолженность  \
family_status                                                          
в разводе                                   1009                  78   
вдовец / вдова                               810                  59   
гражданский брак                            3361                 340   
женат / замужем                            10045                 846   
не женат / не замужем                       2261                 257   

                               %  
family_status                     
в разводе               7.730426  
вдовец / вдова          7.283951  
гражданский брак       10.116037  
женат / замужем         8.422101  
не женат / не замужем  11.366652

### Вывод

Зависимость есть. Чаще других имеют просрочки по кредиту неженатые/незамужние клиенты: среди них 11,3% имели просрочку. Среди женатых/замужних клиентов просрочивших на 3 процентных пункта меньше (8,4%). Ещё меньше вероятность просрочки по кредиту у вдовцов (7,3%) и разведённых клиентов (7,7%). Среди клиентов, состоящих в гражданском браке, не возвращали кредит в срок 10,1%. Таким образом, самыми "ненадежными" для банка клиентами при выдаче кредита являются незамужние/неженатые люди. Самыми "дисциплинированными" можно считать вдов/вдовцов.


<div class="alert alert-success">
<h2> Комментарий ревьюера</h2>

Верно.    
</div>

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [20]:
income_pivot = income_dict.pivot_table(index = ['income_groups'], 
                                columns = 'debt', 
                                values = 'total_income', 
                                aggfunc = 'count')
new = ['Задолженность отсутствует', 'Была задолженность']
income_pivot.set_axis(new, axis = 'columns', inplace = True)
income_pivot['%'] = income_pivot['Была задолженность'] / income_pivot['Задолженность отсутствует'] * 100

income_pivot

Задолженность отсутствует  Была задолженность  \
income_groups                                                                 
Доход 100-200 тысяч рублей                         8518                 855   
Доход более 200 тысяч рублей                       4626                 356   
Доход до 100 тысяч рублей                          3954                 353   

                                      %  
income_groups                            
Доход 100-200 тысяч рублей    10.037568  
Доход более 200 тысяч рублей   7.695633  
Доход до 100 тысяч рублей      8.927668

### Вывод

Зависимость существует. Исходя из полученных данных, можно сказать, что чаще других не возвращают кредит вовремя клиенты с доходом 100-200 тысяч рублей в месяц. Среди них задолженность была 10% клиентов. Состоятельные клиенты с доходом выше 200 тысяч рублей в месяц чаще возвращают кредит в срок (просрочку имеют 7,6%). 91% клиентов с доходом до 100 тысяч рублей возвращают кредит в срок.


<div class="alert alert-success">
<h2> Комментарий ревьюера</h2>

Верно.  
    
Обрати внимание, что наши действия начали повторяться. Если их становится больше двух, рекомендую задуматься о внедрения цилка или же создания универсальной функции, которая будет отображать все необходимые нам расчеты. 
    
</div>

- Как разные цели кредита влияют на его возврат в срок?

In [21]:
purpose_pivot = purpose_dict.pivot_table(index = ['new_purpose'], 
                                columns = 'debt', 
                                values = 'total_income', 
                                aggfunc = 'count')
new = ['Задолженность отсутствует', 'Была задолженность']
purpose_pivot.set_axis(new, axis = 'columns', inplace = True)
purpose_pivot['%'] = purpose_pivot['Была задолженность'] / purpose_pivot['Задолженность отсутствует'] * 100

purpose_pivot

Задолженность отсутствует  Была задолженность          %
new_purpose                                                           
автомобиль                         3495                 368  10.529328
недвижимость                       8847                 717   8.104442
образование                        3237                 334  10.318196
свадьба                            1933                 158   8.173823

### Вывод

Наблюдается следующая зависимость: клиенты, которые берут кредит на автомобиль и образование, не платят его в срок чаще, чем те, кому деньги от банка нужны на операции с недвижимостью и свадьбу. Первые имеют просрочку в 10,5 и 10,3 процентах случаев соответственно. Вторые - в 8,1 и 8,2 процентах случаев.


<div class="alert alert-success">
<h2> Комментарий ревьюера</h2>

Верно.    
</div>

### Шаг 4. Общий вывод

Я получил данные, в которых содержались артефакты, мешавшие корректной работе с ними (отрицательные значения там, где их не должно быть, отсутствующие значения). Значения в столбцах с целями нуждались в лемматизации для дальнейшей работы. После работы с пропусками, удаления дубликатов и лемматизации я смог категоризировать данные, создав 4 словаря для ответа на 4 вопроса из задания. В итоге я выяснил, что возврат кредита в срок в большей или меньшей степени зависит от различных факторов: семейного статуса клиента, наличия у него детей, размера дохода и цели кредита.

<div class="alert alert-success">
<h2> Комментарий ревьюера</h2>

Выводы понятны и логичны. Рекомендую разбавлять его полученными фактами. Таким образом он будет выглядеть конкретнее.
    
Обрати пожалуйста внимание, что финальный вывод и есть главный результат твоей работы. Стоит писать его подробно по результатам проведенной работы. Также можно расписать все, что было сделано в работе.
   
В целом сам проект выполнен на хорошем уровне! На протяжении всей работы чувствуется глубина проработки задачи и это безусловный плюс для нас. Было использовано большое количество методов, которые будут помогать тебе в дальнейших проектах.
    
Не стоит забывать про оформление своей работы, это важный такой же важный этап. Рекомендую «не стесняться» использовать ячейки типа **markdown** там где нам необходимо отобразить свои размышления, в том числе с использованием дополнительной стилизации.
 
Успехов тебе в новых проектах!
    
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера</h2>

Рекомендую посмотреть на дополнительную стилизацию ячеек типа **markdown**.
    
**Жирный**
*Курсив*
    
--- 

# Заголовок первого уровня

--- 
    
## Заголовок первого уровня
  
--- 
Списки:
    
- один 
- два
- три
   
---
    
1. Раз
2. Раз раз
3. Раз два три
    
--- 
    
Отображение `переменных`
    
    
</div>

### Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [ ]  файл изучен;
- [ ]  определены пропущенные значения;
- [ ]  заполнены пропущенные значения;
- [ ]  есть пояснение, какие пропущенные значения обнаружены;
- [ ]  описаны возможные причины появления пропусков в данных;
- [ ]  объяснено, по какому принципу заполнены пропуски;
- [ ]  заменен вещественный тип данных на целочисленный;
- [ ]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [ ]  удалены дубликаты;
- [ ]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [ ]  описаны возможные причины появления дубликатов в данных;
- [ ]  выделены леммы в значениях столбца с целями получения кредита;
- [ ]  описан процесс лемматизации;
- [ ]  данные категоризированы;
- [ ]  есть объяснение принципа категоризации данных;
- [ ]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [ ]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [ ]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [ ]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [ ]  в каждом этапе есть выводы;
- [ ]  есть общий вывод.